client/state/push-notifications/actions.js

```diff
@@ -0,0 +1,324 @@
+								debug( 'Error while unsubscribing', err );
+							} )
+						;
+					} )
+					.catch( err => {
+						debug( 'Error getting subscription to deactivate', err );
+
+						// @TODO is this correct behavior?
+						dispatch( receiveSubscription( null ) );
+					} )
+				;
+			} );
+	};
+}
+
+export function receivePermissionState( permission ) {
+	return dispatch => {
+		if ( permission === 'granted' ) {
+			dispatch( authorize() );
+			return;
+		}
+
+		if ( permission === 'denied' ) {
+			dispatch( block() );
+			return;
+		}
+
+		dispatch( mustPrompt() );
+		dispatch( toggleEnabled() );
+	};
+}
+
+export function fetchPushManagerSubscription() {
+	return dispatch => {
+		window.navigator.serviceWorker.ready
+			.then( ( serviceWorkerRegistration ) => {
+				serviceWorkerRegistration.pushManager.getSubscription()
+					.then( pushSubscription => {
+						dispatch( receiveSubscription( pushSubscription ) );
+						dispatch( sendSubscriptionToWPCOM() );
+					} )
+					.catch( err => debug( 'Error getting subscription', err ) )
+				;
+			} )
+			.catch( err => debug( 'Error fetching push manager subscription', err )	)
+		;
+	};
+}
+
+export function sendSubscriptionToWPCOM() {
+	return ( dispatch, getState ) => {
+		const state = getState();
+		const lastUpdated = getLastUpdated( state );
+		debug( 'Subscription last updated: ' + lastUpdated );
+
+		let age;
+
+		if ( lastUpdated ) {
+			age = moment().diff( moment( lastUpdated ), 'days' );
+			if ( age < DAYS_BEFORE_FORCING_REGISTRATION_REFRESH ) {
+				debug( 'Subscription did not need updating.', age );
+				return;
+			}
+		}
+
+		debug( 'Subscription needed updating.', age );
+
+		const sub = getSavedSubscription( state );
+		if ( ! sub ) {
+			debug( 'No subscription to send to WPCOM' );
+			// @TODO dispatch something :)
+			return;
+		}
+		debug( 'Sending subscription to WPCOM', sub );
+
+		wpcom.undocumented().registerDevice( JSON.stringify( sub ), 'browser', 'Browser' )
+			.then( data => dispatch( {
+				type: PUSH_NOTIFICATIONS_RECEIVE_REGISTER_DEVICE,
+				data
+			} ) )
+			.catch( err => debug( 'Couldn\'t register device', err ) )
+		;
+	};
+}
+
+export function activateSubscription() {
+	return dispatch => {
+		window.navigator.serviceWorker.ready
+			.then( serviceWorkerRegistration => {
+				serviceWorkerRegistration.pushManager.subscribe( { userVisibleOnly: true } )
+					.then( subscription => {
+						dispatch( receiveSubscription( subscription ) );
+						dispatch( checkPermissionsState() );
+					} )
+					.catch( err => {
+						debug( 'Error receiving subscription', err );
+						dispatch( block() );
+					} )
+				;
+			} )
+			.catch( err => debug( 'Error activating subscription', err )	)
+		;
+	};
+}
+
+export function unregisterDevice() {
+	return ( dispatch, getState ) => {
+		const deviceId = getDeviceId( getState() );
+		if ( ! deviceId ) {
+			debug( 'Couldn\'t unregister device. Unknown device ID' );
+			dispatch( receiveUnregisterDevice( {} ) );
```

Comment:
Handle a null argument in `receiveUnregisterDevice` so I don't have to pass objects here